### PR TITLE
Feature: redesign create a diagram from template

### LINKED ISSUE #282: Project page (index page): redesign form to create a diagram from a template
## Description

On the project page (index page), change the dialog to create a diagram by a drawer.

To implements: 
- Drawer global design
- Search bar
- plugin list
- diagram list

:warning: For diagrams list, is not mention in the mockup, but put them in a scroll panel.

To implements but not on the mockup (see #226, for more details):
- input for all the possible folders
- input for filename
- Some label to inform user of the new diagrams

:warning: For all missing things, feel free to redesign as your wish the mockup!

Not to implements
- All tags

## Mockup

image

### PR DESCRIPTION
**Pull request by Zorin95670**
Resolved #282

### CODE REVIEW

File: cypress/e2e/ModelsPage/FilterTemplate.feature [hunk 1]
```diff
@@ -0,0 +1,38 @@
+Feature: Test models page: template filter
+
+  Scenario: Should filter templates by tags and search text
```
Comment (justinehell) [inline cypress/e2e/ModelsPage/FilterTemplate.feature:3]: Plugin selection is considered as tag ?

File: changelog.md [hunk 2]
```diff
@@ -13,6 +13,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   * Focus movement on all diagrams.
 * Index page:
   * Add a search bar to filter diagrams.
+  * Add drawer to select template of a diagram.
```
Comment (justinehell) [inline changelog.md:16]: Suggestion : `Add drawer to create diagram from a template.`

File: src/i18n/en-US/index.js [hunk 3]
```diff
@@ -35,6 +36,12 @@ export default {
       },
     },
     models: {
+      add: {
+        button: {
+          name: 'Add a diagram',
+          title: 'Add a diagram from template',
```
Comment (k-y0u) [inline src/i18n/en-US/index.js:42]: from A template

File: src/i18n/en-US/index.js [hunk 4]
```diff
@@ -264,6 +271,13 @@ export default {
         plugin: 'Plugin',
         actionsTitle: 'Tools',
       },
+      drawer: {
+        title: 'Add a template of a model',
+        search: 'Search a template of a model',
+        list: 'Models templates',
+        select: 'Plugins...',
+        empty: 'No templates available',
```
Comment (k-y0u) [inline src/i18n/en-US/index.js:279]: No template (singular)

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                :key="template.key"
+                class=""
+                :class="[
+                  'template-card-item',
+                  'row', 'q-mb-md',
+                  'justify-center',
+                  'items-center',
+                  'q-pa-sm',
+                  'rounded-borders',
+                  selectedTemplate === template.key ? 'selected' : null,
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:98]: { selected: selectedTemplate === template.key },

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                ]"
+                @click="selectedTemplate = selectedTemplate === template.key
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:100]: too much code in html

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+/**
+ * Update templates list.
+ */
+function updateTemplates() {
+  templates.value = allTemplates
+    .filter(({ type }) => searchText(type, searchTemplateText.value))
+    .filter(({ plugin }) => {
+      if (!selectedPlugin.value) {
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:188]: it's ugly

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+/**
+ * Reset all inputs.
+ */
+function reset() {
+  searchTemplateText.value = null;
+  selectedTemplate.value = null;
+  selectedPlugin.value = null;
+  updateTemplates();
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:203]: jump a line before

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+ * Open or close drawer on event.
+ * @param {Object} event - Event.
+ * @param {String} event.key - Identifier of drawer.
+ * @param {String} event.type - Action type, 'open' or 'close'.
+ */
+function onDrawerEvent({ key, type }) {
+  if (key !== 'CreateDiagram') {
+    return;
+  }
+  const isOpen = type === 'open';
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:216]: jump a line before

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+  if (isOpen) {
+    reset();
+  }
+  emit('update:modelValue', isOpen);
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:221]: jump a line before

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+onMounted(async () => {
+  drawerEventSubscription = DrawerEvent.subscribe(onDrawerEvent);
+  allTemplates = await getTemplatesByType('model');
+  updateTemplates();
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:227]: jump a line before

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+  &.selected {
+    border: 1px solid $accent;
+  }
+}
+.plugins-select, .search-bar {
+  max-width: 250px;
+  width: 250px;
+}
+
```
Comment (k-y0u) [inline src/components/drawer/CreateDiagramDrawer.vue:265]: delete

File: cypress/e2e/ModelsPage/FilterTemplate.feature [hunk 13]
```diff
@@ -0,0 +1,38 @@
+Feature: Test models page: template filter
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/FilterTemplate.feature:1]: filter template

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/layouts/ProjectLayout.vue [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+</template>
+
+<script setup>
+import CreateDiagramDrawer from 'components/drawer/CreateDiagramDrawer.vue';
+import NavigationBar from 'components/NavigationBar.vue';
+import { useRoute } from 'vue-router';
+import { computed, ref } from 'vue';
+
+const route = useRoute();
+const projectName = computed(() => route.params.projectName);
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/layouts/ProjectLayout.vue:27]: done

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+        </q-item-section>
+      </q-item>
+      <q-item class="list-item">
+        <q-item-section>
+          <q-select
+            v-model="selectedPlugin"
+            outlined
+            dense
+            clearable
+            class="plugins-select"
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:63]: done

File: src/components/drawer/CreateDiagramDrawer.vue [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+        </q-item-section>
+      </q-item>
+      <q-item class="list-item-shrink">
+        <q-item-section>
+          <q-scroll-area style="height: 100%">
+            <div class="template-card-container row items-center wrap">
+              <div
+                v-for="template in templates"
+                :key="template.key"
+                class=""
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:90]: done

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 8 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:203]: done

File: tests/unit/components/drawer/CreateDiagramDrawer.spec.js [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+      expect(wrapper.vm.templates).toEqual([{
+        type: 'test1',
+        plugin: 'terrator-plugin',
+      }, {
+        type: 'test2',
+        plugin: 'githubator-plugin',
+      }]);
+    });
+
+    it('should return only test1 templates', () => {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/components/drawer/CreateDiagramDrawer.spec.js:81]: done

File: tests/unit/components/drawer/CreateDiagramDrawer.spec.js [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+      wrapper.vm.selectedPlugin = null;
+      wrapper.vm.updateTemplates();
+
+      expect(wrapper.vm.templates).toEqual([{
+        type: 'test1',
+        plugin: 'terrator-plugin',
+      }]);
+    });
+
+    it('should return only githubator templates', () => {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/components/drawer/CreateDiagramDrawer.spec.js:92]: done

File: tests/unit/components/drawer/CreateDiagramDrawer.spec.js [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+  });
+
+  describe('Test function: onDrawerEvent', () => {
+    it('should do nothing on bad key', () => {
+      wrapper.vm.onDrawerEvent({ key: 'bad' });
+
+      expect(wrapper.emitted()).toEqual({});
+    });
+
+    it('should reset inputs on open and emit an event', () => {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/components/drawer/CreateDiagramDrawer.spec.js:134]: done

File: tests/unit/components/drawer/CreateDiagramDrawer.spec.js [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+      expect(wrapper.vm.templates).toEqual([{
+        type: 'test1',
+        plugin: 'terrator-plugin',
+      }, {
+        type: 'test2',
+        plugin: 'githubator-plugin',
+      }]);
+    });
+
+    it('should not reset inputs on close and emit an event', () => {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/components/drawer/CreateDiagramDrawer.spec.js:157]: done

File: cypress/e2e/ModelsPage/FilterTemplate.feature (hunk 1 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/FilterTemplate.feature:3]: nope sorry, done

File: changelog.md (hunk 2 above)
Comment (Zorin95670) [reply changelog.md:16]: done

File: src/i18n/en-US/index.js (hunk 4 above)
Comment (Zorin95670) [reply src/i18n/en-US/index.js:279]: done

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 6 above)
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:100]: done

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 7 above)
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:188]: done

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 8 above)
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:203]: It's remove, so done I guess

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 9 above)
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:216]: done

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 10 above)
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:221]: done

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 11 above)
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:227]: done

File: src/components/drawer/CreateDiagramDrawer.vue (hunk 12 above)
Comment (Zorin95670) [reply src/components/drawer/CreateDiagramDrawer.vue:265]: done

File: cypress/e2e/ModelsPage/FilterTemplate.feature (hunk 13 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/FilterTemplate.feature:1]: done

File: src/layouts/ProjectLayout.vue [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+  width: 100%;
+  height: 100%;
+  background-color: rgba(0, 0, 0, 0.5);
+  z-index: 2001;
+}
+</style>
+
+<style lang="scss">
+.q-drawer:has(.create-diagram-drawer) {
+  width: 43.75% !important;
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/layouts/ProjectLayout.vue:45]: done

**Review by justinehell — APPROVED** [review]

**Review by k-y0u — APPROVED** [review]

**Pull request merged**